File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewGroupHierarchyChangeEventObservable.java
package com.jakewharton.rxbinding2.view;

import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.OnHierarchyChangeListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class ViewGroupHierarchyChangeEventObservable
    extends Observable<ViewGroupHierarchyChangeEvent> {
  private final ViewGroup viewGroup;

  ViewGroupHierarchyChangeEventObservable(ViewGroup viewGroup) {
    this.viewGroup = viewGroup;
  }

  @Override
  protected void subscribeActual(Observer<? super ViewGroupHierarchyChangeEvent> observer) {
    verifyMainThread();
    Listener listener = new Listener(viewGroup, observer);
    observer.onSubscribe(listener);
    viewGroup.setOnHierarchyChangeListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnHierarchyChangeListener {
    private final ViewGroup viewGroup;
    private final Observer<? super ViewGroupHierarchyChangeEvent> observer;

    Listener(ViewGroup viewGroup, Observer<? super ViewGroupHierarchyChangeEvent> observer) {
      this.viewGroup = viewGroup;
      this.observer = observer;
    }

    @Override public void onChildViewAdded(View parent, View child) {
      if (!isDisposed()) {
        observer.onNext(ViewGroupHierarchyChildViewAddEvent.create(viewGroup, child));
      }
    }

    @Override public void onChildViewRemoved(View parent, View child) {
      if (!isDisposed()) {
        observer.onNext(ViewGroupHierarchyChildViewRemoveEvent.create(viewGroup, child));
      }
    }

    @Override protected void onDispose() {
      viewGroup.setOnHierarchyChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewGroupHierarchyChangeEventObservable.kt
@file:JvmName("RxViewGroup")
@file:JvmMultifileClass

package com.jakewharton.rxbinding2.view

import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.OnHierarchyChangeListener
import androidx.annotation.CheckResult
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding2.internal.checkMainThread

/**
 * Create an observable of hierarchy change events for `viewGroup`.
 *
 * *Warning:* The created observable keeps a strong reference to `viewGroup`.
 * Unsubscribe to free this reference.
 */
@CheckResult
fun ViewGroup.changeEvents(): Observable<ViewGroupHierarchyChangeEvent> {
  return ViewGroupHierarchyChangeEventObservable(this)
}

private class ViewGroupHierarchyChangeEventObservable(
  private val viewGroup: ViewGroup
) : Observable<ViewGroupHierarchyChangeEvent>() {

  override fun subscribeActual(observer: Observer<in ViewGroupHierarchyChangeEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(viewGroup, observer)
    observer.onSubscribe(listener)
    viewGroup.setOnHierarchyChangeListener(listener)
  }

  private class Listener(
    private val viewGroup: ViewGroup,
    private val observer: Observer<in ViewGroupHierarchyChangeEvent>
  ) : MainThreadDisposable(), OnHierarchyChangeListener {

    override fun onChildViewAdded(parent: View, child: View) {
      if (!isDisposed) {
        observer.onNext(ViewGroupHierarchyChildViewAddEvent(viewGroup, child))
      }
    }

    override fun onChildViewRemoved(parent: View, child: View) {
      if (!isDisposed) {
        observer.onNext(ViewGroupHierarchyChildViewRemoveEvent(viewGroup, child))
      }
    }

    override fun onDispose() {
      viewGroup.setOnHierarchyChangeListener(null)
    }
  }
}
